frontend/utils/distribution/monthly-event-parser.js

```diff
@@ -0,0 +1,199 @@
+  const paymentsDistribution = function (groupMembers, minCome) {
+    const groupIncomes = groupMembers.map((user) => {
+      return {
+        name: user.name,
+        amount: minCome + user.cyclicalIncomeIntegral
+      }
+    })
+    return incomeDistribution(groupIncomes, minCome)
+  }
+  // Loop through the events, pro-rating each user's monthly pledges/needs:
+  for (const event of distributionEvents) {
+    if (event.type === 'startCycleEvent') {
+      monthlyDistribution = paymentsDistribution(proRateHaveNeeds(groupMembers), minCome)
+
+      // Check if it is the last event (the next month after monthstamps cycle event), or if the
+      // final distribution should be adjusted, anyway:
+      monthlyDistribution = subtractDistributions(monthlyDistribution, completedMonthlyPayments)
+
+      const overPayments = monthlyDistribution.filter((p) => {
+        return p.amount < 0
+      }).map((p) => {
+        p.amount = Math.abs(p.amount)
+        return p
+      })
+
+      const latePayments = monthlyDistribution.filter((p) => {
+        return p.amount > 0
+      })
+
+      lastStartCycleEvent = event
+      lastStartCycleEvent.data.latePayments = redistributeOverToLatePayments(overPayments, latePayments)
+
+      // Reset the income distribution calcuulation at the start of each cycle...
+      for (const member of groupMembers) {
+        member.cyclicalIncomeVariable = 0
+        member.cyclicalIncomeIntegral = 0
+      }
+
+      // Unadjust last
+      if (!adjusted) {
+        lastStartCycleEvent.data.latePayments = reduceDistribution(addDistributions(completedMonthlyPayments, lastStartCycleEvent.data.latePayments))
+      }
+      completedMonthlyPayments = [] // and the monthly payments, too...
+    } else if (event.type === 'haveNeedEvent') {
+      const oldUser = getUser(event.data.name)
+      const cyclesIntoMonth = event.data.cycle % 1
+      if (oldUser) {
+        proRateHaveNeeds([oldUser], cyclesIntoMonth)
+        oldUser.haveNeed = event.data.haveNeed
+      } else {
+        // Add the user who declared their income to our groupMembers list variable
+        groupMembers.push({
+          name: event.data.name,
+          haveNeed: event.data.haveNeed,
+          cyclicalIncomeVariable: event.data.cycle % 1,
+          cyclicalIncomeIntegral: 0
+        })
+      }
+    } else if (event.type === 'paymentEvent') {
+      completedMonthlyPayments.push({
+        from: event.data.from,
+        to: event.data.to,
+        amount: event.data.amount
+      })
+    } else if (event.type === 'userExitsGroupEvent') {
+      groupMembers = groupMembers.filter((v) => { return v.name !== event.data.name })
+    }
+  }
+
+  return lastStartCycleEvent.data.latePayments.filter((payment) => {
+    return payment.from !== payment.to // This happens when a haver switches to being a needer; remove neutral distribution payments.
+  }) // TODO: return late-payments as well.
+}
+
+export default parseMonthlyDistributionFromEvents
```

Comment:
In this file, can you:

- Remove as many `any`s as you can?
- Simplify and dry the code as much as you can? (E.g. reduce the linecount)?

Reply:
> Remove as many anys as you can?

Done.

> Simplify and dry the code as much as you can? (E.g. reduce the linecount)?

See response to next comment. It's a start.